torch/utils/cpp_extension.py

```diff
@@ -0,0 +1,252 @@
+
+    if build_directory is None:
+        build_directory = _get_build_directory(name, verbose)
+
+    build_file_path = os.path.join(build_directory, 'build.ninja')
+    if verbose:
+        print('Emitting ninja build file {}...'.format(build_file_path))
+    # NOTE: Emitting a new ninja build file does not cause re-compilation if
+    # the sources did not change, so it's ok to re-emit (and it's fast).
+    _write_ninja_file(build_file_path, name, sources, extra_cflags or [],
+                      extra_ldflags or [], extra_include_paths or [])
+
+    if verbose:
+        print('Building extension module {}...'.format(name))
+    _build_extension_module(name, build_directory)
+
+    if verbose:
+        print('Loading extension module {}...'.format(name))
+    return _import_module_from_library(name, build_directory)
+
+
+def verify_ninja_availability():
+    try:
+        check = 'ninja --version'.split()
+        subprocess.check_call(check, stdout=open(os.devnull))
+    except OSError:
+        raise RuntimeError("Ninja is required to load C++ extensions")
+
+
+def _get_build_directory(name, verbose):
+    root_extensions_directory = os.environ.get('TORCH_EXTENSIONS_DIR')
+    if root_extensions_directory is None:
+        # tempfile.gettempdir() will be /tmp on UNIX and \TEMP on Windows.
+        root_extensions_directory = os.path.join(tempfile.gettempdir(),
+                                                 'torch_extensions')
+
+    if verbose:
+        print('Using {} as PyTorch extensions root...'.format(
+            root_extensions_directory))
+
+    build_directory = os.path.join(root_extensions_directory, name)
+    if not os.path.exists(build_directory):
+        if verbose:
+            print('Creating extension directory {}...'.format(build_directory))
+        # This is like mkdir -p, i.e. will also create parent directories.
+        os.makedirs(build_directory)
+
+    return build_directory
+
+
+def _build_extension_module(name, build_directory):
+    try:
+        subprocess.check_output(
+            ['ninja', '-v'], stderr=subprocess.STDOUT, cwd=build_directory)
+    except subprocess.CalledProcessError:
+        # Python 2 and 3 compatible way of getting the error object.
+        _, error, _ = sys.exc_info()
+        # error.output contains the stdout and stderr of the build attempt.
+        raise RuntimeError("Error building extension '{}': {}".format(
+            name, error.output.decode()))
+
+
+def _import_module_from_library(module_name, path):
+    # https://stackoverflow.com/questions/67631/how-to-import-a-module-given-the-full-path
+    file, path, description = imp.find_module(module_name, [path])
+    # Close the .so file after load.
+    with file:
+        return imp.load_module(module_name, file, path, description)
+
+
+def _write_ninja_file(path, name, sources, extra_cflags, extra_ldflags,
+                      extra_include_paths):
+    # Version 1.3 is required for the `deps` directive.
```

Comment:
@zdevito I've decided to rewrite this without the ninja writer so that we don't have to depend on the ninja Python module, i.e. it's no longer required to have `pip install ninja` (which installs the `ninja` Python module *and* the `ninja` command line tool/build system), it's sufficient to have the `ninja` build system, which you get from `apt-get install ninja-build`, `brew install ninja`, `conda install ninja` etc.
It also removes the pain of having to install `ninja` with `pip` in the build system which has been a huge pain